### ISSUE TITLE
What do you guys do/use in working on the project, especially during testing?

### ISSUE THREAD

**Issue by seanballais** [darlinghq/darling#431]
I am interested in contributing to the project in my free time. I read that working on this project requires installing an experimental (and most likely unstable) kernel module after building the project. I am wary of doing so as it might soft-brick my machine. I could use a VM but doing so seems to add additional steps to testing.

Is there an alternative to using a VM or directly installing the kernel module, such as somehow isolating the module (maybe with some sort of a container)?

**Comment by seanballais** [comment]
I see. Thanks!

Perhaps using an automation tool that'll compile and install the module in the test machine/VM when new commits are pushed would speed the process.

Why does Darling need a kernel module in the first place?

**Comment by ahyattdev** [comment]
@Guy1524 it also handles Mach IPC, Mach-O loading, and similar tasks.

**Comment by stek29** [comment]
@seanballais because mach ipc exposes a lot of stuff which just can't be accessed outside of kernel (to provide this mach api)

**Comment by LubosD** [comment]
@seanballais Yes, there would be a huge overhead. If you take a look at Wine, you'll frequently see `wineserver` being blamed for performance issues, especially if an application needlessly uses mutexes (which are kernel objects on Windows) instead of critical sections (which are like futexes in the Unix world).

The interaction between a Mac application and Mach IPC facilities is even heavier than between an ordinary Windows app and wineserver.

Furthermore, Mach kernel servers do things that no userspace process could easily do: for instance, they allow reading/writing/(de)allocating other process' memory.
